File: web/lib/llm.ts
/**
 * Fever HQ — Anthropic API wrapper.
 *
 * Two upgrades over v1:
 *   1) Date awareness is computed in America/Chicago (Central), not UTC. Fixes
 *      the bug where the bot would claim it was "tomorrow" after 7pm Central.
 *   2) Server-side web_search tool is enabled. The model can answer live
 *      questions about Caitlin Clark and the Fever — scores, starting lineups,
 *      stat lines, injury news — by searching official sources at reply time.
 *
 * No new env vars required. ANTHROPIC_API_KEY covers web_search billing.
 */
import Anthropic from "@anthropic-ai/sdk";
import { schedule, upcomingGames, pastGames } from "./schedule";

const MODEL = "claude-haiku-4-5-20251001";

export type ChatMessage = {
  role: "user" | "assistant";
  content: string;
};

// ──────────────────────────────────────────────────────────────────────
// Time helpers — always Central, never UTC.
// ──────────────────────────────────────────────────────────────────────
function todayInCT(): string {
  // YYYY-MM-DD in Chicago time. en-CA gives ISO-style by default.
  return new Intl.DateTimeFormat("en-CA", {
    timeZone: "America/Chicago",
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  }).format(new Date());
}

function nowInCTHuman(): string {
  // "Sunday, June 15, 2026 at 9:42 PM CDT"
  return new Intl.DateTimeFormat("en-US", {
    timeZone: "America/Chicago",
    weekday: "long",
    year: "numeric",
    month: "long",
    day: "numeric",
    hour: "numeric",
    minute: "2-digit",
    hour12: true,
    timeZoneName: "short",
  }).format(new Date());
}

// ──────────────────────────────────────────────────────────────────────
// Schedule knowledge slice — gets injected into the system prompt.
// ──────────────────────────────────────────────────────────────────────
function scheduleContext(): string {
  const today = todayInCT();
  const upcoming = upcomingGames()
    .slice(0, 12)
    .map(({ date, tipoff_local, opponent, home_away, venue, channel }) => ({
      date,
      tipoff_et: tipoff_local,
      opponent,
      home_away,
      venue,
      channel,
    }));
  const recent = pastGames()
    .slice(0, 5)
    .map(({ date, opponent, home_away, channel }) => ({
      date,
      opponent,
      home_away,
      channel,
    }));
  return JSON.stringify(
    {
      team: schedule.team,
      today_in_central: today,
      next_12_games: upcoming,
      last_5_games: recent,
      season_break: "FIBA Women's World Cup pause Aug 31 - Sep 16, 2026",
    },
    null,
    2,
  );
}

// ──────────────────────────────────────────────────────────────────────
// System prompt — voice + constraints + live-data permissions.
// ──────────────────────────────────────────────────────────────────────
function systemPrompt(): string {
  return `You are Fever HQ, a personal WNBA superfan service. You text with one specific person: a Caitlin Clark and Indiana Fever superfan. She is non-technical. She wants the right info, fast, with personality.

DATE AND TIME
Right now it is ${nowInCTHuman()}. Trust this over anything in your training data. If she asks "is the game tonight" or "what day is it", anchor to this exact date. Never assume the date is tomorrow because of a timezone calculation.

VOICE
- Sports radio energy. Punchy. Confident. Short sentences.
- A little personality, never corporate, never sycophantic.
- Use the words "Fever," "Caitlin," "Aliyah," and player nicknames naturally.
- Emojis sparingly: 🏀 is fine, one per reply max.

CONSTRAINTS
- Web chat replies. Keep them tight: 1 to 3 short sentences for simple questions; up to a short paragraph when you have pulled live data.
- ALWAYS convert game times to CENTRAL TIME (CT). The schedule below lists Eastern; subtract 1 hour and stamp "CT" on every time you quote.
- When she asks "what channel", call out YouTube TV availability for the major broadcast networks. YouTube TV carries: NBA TV, ESPN, ESPN2, ABC, CBS, NBC, ION, USA Network, FX, MeTV. Prime Video, Paramount+, and Peacock are separate streaming services.
- Schedule, channel, and venue questions: answer from the knowledge below first. No need to search the web for those.

LIVE DATA — you have web access via the web_search tool
- For anything live or current, use web_search: tonight's score, did Caitlin start, her stat line, injury reports, trade rumors, recent news, postgame quotes, anything Caitlin Clark or Indiana Fever related that the static schedule below cannot answer.
- Prefer official sources in this order: the Indiana Fever X feed (@IndianaFever), wnba.com, ESPN, The Athletic.
- When you cite live info, briefly name the source in plain English. Examples: "per ESPN", "per the Fever X feed", "per WNBA.com box score". One short attribution is enough.
- If a search returns nothing solid, say so plainly and point her to the Fever app or @IndianaFever to check directly. Never fabricate a score, lineup, or stat.
- Search efficiently: 1 to 2 well-targeted queries per question is usually plenty. Do not burn searches on schedule questions you can already answer from the knowledge below.

KNOWLEDGE
Schedule data follows (Eastern source — convert to CT for the user):
${scheduleContext()}

Reply with just the chat body. No quote marks, no preamble.`;
}

// Anthropic's server-side web search tool. The model decides when to call it;
// Anthropic handles the search itself and returns results into the conversation
// before the model emits its final text. No new env var, no extra plumbing.
const WEB_SEARCH_TOOL = {
  type: "web_search_20250305",
  name: "web_search",
  max_uses: 4,
} as const;

// ──────────────────────────────────────────────────────────────────────
// Streaming chat — yields text deltas only. Citation deltas are ignored
// because the prompt asks the model to name sources in natural language.
// ──────────────────────────────────────────────────────────────────────
export async function* streamReply(
  history: ChatMessage[],
): AsyncIterable<string> {
  const apiKey = process.env.ANTHROPIC_API_KEY;
  if (!apiKey) throw new Error("ANTHROPIC_API_KEY not set");

  const client = new Anthropic({ apiKey });
  const stream = client.messages.stream({
    model: MODEL,
    max_tokens: 800,
    system: systemPrompt(),
    messages: history,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    tools: [WEB_SEARCH_TOOL as any],
  });

  for await (const event of stream) {
    if (
      event.type === "content_block_delta" &&
      event.delta.type === "text_delta"
    ) {
      yield event.delta.text;
    }
  }
}

// Non-streaming variant for one-shot tools (e.g., notification drafting).
export async function generateOneShot(prompt: string): Promise<string> {
  const apiKey = process.env.ANTHROPIC_API_KEY;
  if (!apiKey) throw new Error("ANTHROPIC_API_KEY not set");
  const client = new Anthropic({ apiKey });
  const res = await client.messages.create({
    model: MODEL,
    max_tokens: 800,
    system: systemPrompt(),
    messages: [{ role: "user", content: prompt }],
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    tools: [WEB_SEARCH_TOOL as any],
  });
  return res.content
    .filter((b) => b.type === "text")
    .map((b) => (b as { text: string }).text)
    .join("");
}

File: src/llm.py
"""
Fever HQ — Anthropic API wrapper for bot replies.

Loads schedule.json as live knowledge, builds a sports-radio system prompt,
calls Haiku 4.5, returns text trimmed to SMS-friendly length.

Required env vars:
  ANTHROPIC_API_KEY

Honest scope note:
  v1 answers schedule-based questions ("when's next game?", "who plays
  Saturday?", "what channel?") from schedule.json. Live "did Caitlin score 30
  tonight?" / "who started?" questions require fresh data — added as tool-use
  in v2. v1 explicitly tells the user when it doesn't have live data and
  points them to where they CAN check.
"""
from __future__ import annotations
import os
import json
from datetime import date, datetime
from pathlib import Path
from typing import Optional


MODEL = "claude-haiku-4-5-20251001"
SCHEDULE_PATH = Path(__file__).parent / "schedule.json"

SYSTEM_PROMPT_TEMPLATE = """\
You are Fever HQ, a personal WNBA superfan service texting one specific
person: a die-hard Caitlin Clark / Indiana Fever fan. She is not technical.
She wants the right info, fast, with personality.

VOICE
- Sports radio energy. Punchy. Confident. Short sentences.
- A little personality, never sycophantic, never corporate.
- Use the words "Fever," "Caitlin," "Aliyah," and player nicknames naturally.
- Emojis sparingly: maybe one per message. 🏀 is fine. Skip the rest.

CONSTRAINTS
- You are sending SMS. Hard cap: 300 characters. Aim for 1-2 segments.
- Never invent facts. If you don't know — say so plainly and tell her where
  to check (the Fever app, ESPN, the team's X account).
- ALWAYS convert game times to CENTRAL TIME (CT) when replying. The schedule
  data below is listed in Eastern Time (broadcaster convention), but the user
  is on Central — subtract 1 hour from any time you quote. Always end times
  with "CT" so she knows which zone you mean.
- You can answer schedule-based questions from your knowledge below.
- You CANNOT answer live questions (current scores, who scored what tonight,
  injury news from the last hour) — admit it cleanly.

KNOWLEDGE: 2026 Indiana Fever schedule
Today's date: {today}
Schedule data (all times below are EASTERN — convert to CT before showing user):
{schedule_json}

FALLBACK PHRASES (use freely)
- For live questions: "I don't have live game info — pull up the Fever app
  for the latest. I'll text you 15 mins before tipoff every game."
- For unknown players: "Not sure on that one — Fever roster updates I don't
  catch live."
- For off-topic: stay on Fever. Politely redirect.

Reply ONLY with the SMS body. No quote marks, no preamble.
"""


def _load_schedule_summary() -> str:
    data = json.loads(SCHEDULE_PATH.read_text())
    today = date.today()
    # Trim to upcoming + last 3 games — keep prompt small
    upcoming = []
    recent = []
    for g in data["games"]:
        gdate = datetime.strptime(g["date"], "%Y-%m-%d").date()
        if gdate >= today:
            upcoming.append(g)
        else:
            recent.append(g)
    payload = {
        "team": data["team"],
        "next_10_games": upcoming[:10],
        "last_3_games": recent[-3:] if recent else [],
        "season_break": "FIBA Women's World Cup pause Aug 31 - Sep 16, 2026",
    }
    return json.dumps(payload, indent=2)


def reply(user_message: str, *, dry_run: bool = False) -> str:
    """Generate a bot reply for an inbound text.

    Returns SMS-ready text. Trims to 300 chars even if model exceeds.
    """
    user_message = (user_message or "").strip()
    if not user_message:
        return "🏀 Hey! Text me 'next' for the next game or 'when' for tipoff time."

    system_prompt = SYSTEM_PROMPT_TEMPLATE.format(
        today=date.today().isoformat(),
        schedule_json=_load_schedule_summary(),
    )

    if dry_run:
        # Don't hit the API. Useful for unit tests.
        return f"[dry_run] would have generated reply for: {user_message[:80]}"

    api_key = os.environ.get("ANTHROPIC_API_KEY")
    if not api_key:
        raise RuntimeError("ANTHROPIC_API_KEY env var not set")

    try:
        from anthropic import Anthropic
    except ImportError as e:
        raise RuntimeError("anthropic package not installed") from e

    client = Anthropic(api_key=api_key)
    response = client.messages.create(
        model=MODEL,
        max_tokens=256,
        system=system_prompt,
        messages=[{"role": "user", "content": user_message}],
    )
    text = "".join(
        block.text for block in response.content if hasattr(block, "text")
    ).strip()

    # Hard cap to 300 chars — never let the model bust the segment cap
    if len(text) > 300:
        text = text[:297].rstrip() + "..."
    return text
